File: firmware/test/test_imu_types.py
#!/usr/bin/env python3
"""
test_imu_types.py — Host-side unit tests for imu_types.h pure logic

Since this Windows machine has no native g++/gcc, we validate the pure math
from imu_types.h (scale tables, rate divisors, FIFO parsing, timestamp
interpolation) via Python. The C++ code is verified to compile (including
static_assert) via `pio run -e left`.

The math here mirrors imu_types.h exactly — if these tests pass AND pio build
passes, the logic is correct.

Run: python test/test_imu_types.py
"""

import struct
import sys
import unittest

# ── Constants from imu_types.h ────────────────────────────────────────────────

FIFO_SAMPLE_BYTES = 12
FIFO_CAPACITY_BYTES = 512
SAMPLE_QUEUE_LEN = 64
MIN_SAMPLE_RATE_HZ = 50
MAX_SAMPLE_RATE_HZ = 200
DEFAULT_SAMPLE_RATE_HZ = 100

# ── Scale tables (from imu_types.h) ───────────────────────────────────────────

ACCEL_SCALE_TABLE = [
    2.0 / 32768.0,    # ±2g
    4.0 / 32768.0,    # ±4g
    8.0 / 32768.0,    # ±8g
    16.0 / 32768.0,   # ±16g
]

GYRO_SCALE_TABLE = [
    250.0 / 32768.0,    # ±250 dps
    500.0 / 32768.0,    # ±500 dps
    1000.0 / 32768.0,   # ±1000 dps
    2000.0 / 32768.0,   # ±2000 dps
]

# ── Pure functions (mirrors of imu_types.h inline functions) ──────────────────

def accel_scale(range_id):
    return ACCEL_SCALE_TABLE[range_id]

def gyro_scale(range_id):
    return GYRO_SCALE_TABLE[range_id]

def is_valid_rate(rate_hz):
    return rate_hz in (50, 100, 200)

def sample_rate_divisor(rate_hz):
    if not is_valid_rate(rate_hz):
        return 0xFFFF
    return 1000 // rate_hz - 1

def fifo_overflowed(fifo_bytes):
    return fifo_bytes >= FIFO_CAPACITY_BYTES

def parse_fifo_sample(raw_bytes):
    """Parse 12 bytes (big-endian int16) → (ax, ay, az, gx, gy, gz)."""
    ax = struct.unpack('>h', raw_bytes[0:2])[0]
    ay = struct.unpack('>h', raw_bytes[2:4])[0]
    az = struct.unpack('>h', raw_bytes[4:6])[0]
    gx = struct.unpack('>h', raw_bytes[6:8])[0]
    gy = struct.unpack('>h', raw_bytes[8:10])[0]
    gz = struct.unpack('>h', raw_bytes[10:12])[0]
    return (ax, ay, az, gx, gy, gz)

def interpolate_timestamp(drain_us, rate_hz, n_samples, sample_index):
    """Interpolate device timestamp for sample at index i (0=oldest)."""
    interval_us = 1000000 // rate_hz
    offset_us = (n_samples - 1 - sample_index) * interval_us
    return drain_us - offset_us

# ── ImuSample struct size (BLE protocol: 20 bytes) ────────────────────────────
# uint32 seq (4) + uint32 t_device_us (4) + 6×int16 (12) = 20 bytes
IMU_SAMPLE_SIZE = 4 + 4 + 6 * 2

# ── Test cases ────────────────────────────────────────────────────────────────

class TestImuSampleStruct(unittest.TestCase):
    """Journey: BLE packet must be exactly 20 bytes (docs/ble-protocol.md §2.1)"""

    def test_struct_size_is_20_bytes(self):
        self.assertEqual(IMU_SAMPLE_SIZE, 20)

    def test_field_offsets(self):
        # seq=0, t_device_us=4, ax=8, ay=10, az=12, gx=14, gy=16, gz=18
        offsets = {
            'seq': 0, 't_device_us': 4,
            'ax': 8, 'ay': 10, 'az': 12,
            'gx': 14, 'gy': 16, 'gz': 18,
        }
        expected_size = max(offsets.values()) + 2  # last field is int16
        self.assertEqual(expected_size, 20)


class TestScaleTables(unittest.TestCase):
    """Journey: app must convert raw LSB → physical units using correct scale"""

    def test_accel_scale_g2(self):
        self.assertAlmostEqual(accel_scale(0), 2.0 / 32768.0, places=7)

    def test_accel_scale_g4(self):
        self.assertAlmostEqual(accel_scale(1), 4.0 / 32768.0, places=7)

    def test_accel_scale_g8(self):
        self.assertAlmostEqual(accel_scale(2), 8.0 / 32768.0, places=7)

    def test_accel_scale_g16(self):
        self.assertAlmostEqual(accel_scale(3), 16.0 / 32768.0, places=7)

    def test_gyro_scale_dps250(self):
        self.assertAlmostEqual(gyro_scale(0), 250.0 / 32768.0, places=7)

    def test_gyro_scale_dps2000(self):
        self.assertAlmostEqual(gyro_scale(3), 2000.0 / 32768.0, places=7)


class TestRateValidation(unittest.TestCase):
    """Journey: sample rate must be exactly 50/100/200 Hz (not arbitrary)"""

    def test_valid_rates(self):
        self.assertTrue(is_valid_rate(50))
        self.assertTrue(is_valid_rate(100))
        self.assertTrue(is_valid_rate(200))

    def test_invalid_rates(self):
        self.assertFalse(is_valid_rate(0))
        self.assertFalse(is_valid_rate(75))
        self.assertFalse(is_valid_rate(150))
        self.assertFalse(is_valid_rate(201))
        self.assertFalse(is_valid_rate(500))

    def test_divisor_50hz(self):
        self.assertEqual(sample_rate_divisor(50), 19)    # 1000/50 - 1

    def test_divisor_100hz(self):
        self.assertEqual(sample_rate_divisor(100), 9)    # 1000/100 - 1

    def test_divisor_200hz(self):
        self.assertEqual(sample_rate_divisor(200), 4)    # 1000/200 - 1

    def test_divisor_invalid_returns_sentinel(self):
        self.assertEqual(sample_rate_divisor(75), 0xFFFF)
        self.assertEqual(sample_rate_divisor(0), 0xFFFF)


class TestFifoOverflow(unittest.TestCase):
    """Journey: FIFO overflow must be detected so data loss is known"""

    def test_not_overflowed_under_capacity(self):
        self.assertFalse(fifo_overflowed(0))
        self.assertFalse(fifo_overflowed(120))    # 10 samples
        self.assertFalse(fifo_overflowed(504))    # 42 samples
        self.assertFalse(fifo_overflowed(511))    # just under

    def test_overflowed_at_capacity(self):
        self.assertTrue(fifo_overflowed(512))     # exactly capacity
        self.assertTrue(fifo_overflowed(1024))    # way over


class TestFifoParsing(unittest.TestCase):
    """Journey: FIFO data is big-endian int16, must parse correctly"""

    def test_positive_values(self):
        raw = struct.pack('>hhhhhh', 100, 200, 300, 10, 20, 30)
        ax, ay, az, gx, gy, gz = parse_fifo_sample(raw)
        self.assertEqual((ax, ay, az, gx, gy, gz), (100, 200, 300, 10, 20, 30))

    def test_negative_values(self):
        raw = struct.pack('>hhhhhh', -100, -200, -300, -10, -20, -30)
        ax, ay, az, gx, gy, gz = parse_fifo_sample(raw)
        self.assertEqual((ax, ay, az, gx, gy, gz), (-100, -200, -300, -10, -20, -30))

    def test_max_min_values(self):
        raw = struct.pack('>hhhhhh', 32767, -32768, 0, 32767, -32768, 0)
        ax, ay, az, gx, gy, gz = parse_fifo_sample(raw)
        self.assertEqual((ax, ay, az, gx, gy, gz), (32767, -32768, 0, 32767, -32768, 0))


class TestTimestampInterpolation(unittest.TestCase):
    """Journey: each sample in a batch must have a distinct, accurate timestamp
    so clock sync (#7) can align L/R data. The bug was: all samples got the
    same drain-time timestamp. The fix: interpolate backwards from drain time.
    """

    def test_single_sample_equals_drain_time(self):
        self.assertEqual(interpolate_timestamp(1000000, 100, 1, 0), 1000000)

    def test_oldest_in_batch(self):
        # 5 samples @ 100 Hz (interval=10000us), oldest (i=0):
        # drain - (5-1-0)*10000 = drain - 40000
        drain = 5000000
        self.assertEqual(interpolate_timestamp(drain, 100, 5, 0), drain - 40000)

    def test_newest_in_batch(self):
        # 5 samples @ 100 Hz, newest (i=4): drain - 0 = drain
        drain = 5000000
        self.assertEqual(interpolate_timestamp(drain, 100, 5, 4), drain)

    def test_middle_in_batch(self):
        # 5 samples @ 100 Hz, middle (i=2): drain - 20000
        drain = 5000000
        self.assertEqual(interpolate_timestamp(drain, 100, 5, 2), drain - 20000)

    def test_200hz_interval(self):
        # 3 samples @ 200 Hz (interval=5000us), oldest: drain - 10000
        drain = 2000000
        self.assertEqual(interpolate_timestamp(drain, 200, 3, 0), drain - 10000)

    def test_50hz_interval(self):
        # 2 samples @ 50 Hz (interval=20000us), oldest: drain - 20000
        drain = 1000000
        self.assertEqual(interpolate_timestamp(drain, 50, 2, 0), drain - 20000)

    def test_timestamps_are_distinct(self):
        """The core bug fix: timestamps must all be different"""
        drain = 5000000
        timestamps = [interpolate_timestamp(drain, 100, 5, i) for i in range(5)]
        self.assertEqual(len(set(timestamps)), 5)  # all unique
        # strictly increasing
        for i in range(4):
            self.assertLess(timestamps[i], timestamps[i + 1])

    def test_spacing_matches_rate(self):
        """Consecutive samples must be exactly 1/rate_hz apart"""
        drain = 5000000
        t0 = interpolate_timestamp(drain, 100, 5, 0)
        t1 = interpolate_timestamp(drain, 100, 5, 1)
        self.assertEqual(t1 - t0, 10000)  # 1/100 Hz = 10000 us

    def test_spacing_200hz(self):
        drain = 5000000
        t0 = interpolate_timestamp(drain, 200, 3, 0)
        t1 = interpolate_timestamp(drain, 200, 3, 1)
        self.assertEqual(t1 - t0, 5000)  # 1/200 Hz = 5000 us


if __name__ == '__main__':
    unittest.main(verbosity=2)

File: firmware/test/test_imu_types.cpp
// test/test_imu_types.cpp — Host-side unit tests for imu_types.h pure logic
//
// Tests: struct size, scale tables, rate validation, FIFO overflow detection,
// FIFO byte parsing, timestamp interpolation.
//
// Run: pio test -e native
// These tests run on the host (no ESP32 hardware needed).

#include <unity.h>
#include <cstring>
#include <cmath>
#include "../src/imu_types.h"

using namespace wheelsense;

void setUp(void) {}
void tearDown(void) {}

// ── 1. ImuSample struct size (BLE protocol §2.1: must be 20 bytes) ──────────

void test_imu_sample_is_20_bytes(void) {
    TEST_ASSERT_EQUAL(20, sizeof(ImuSample));
}

void test_imu_sample_field_offsets(void) {
    ImuSample s{};
    TEST_ASSERT_EQUAL(0,  offsetof(ImuSample, seq));
    TEST_ASSERT_EQUAL(4,  offsetof(ImuSample, t_device_us));
    TEST_ASSERT_EQUAL(8,  offsetof(ImuSample, ax));
    TEST_ASSERT_EQUAL(10, offsetof(ImuSample, ay));
    TEST_ASSERT_EQUAL(12, offsetof(ImuSample, az));
    TEST_ASSERT_EQUAL(14, offsetof(ImuSample, gx));
    TEST_ASSERT_EQUAL(16, offsetof(ImuSample, gy));
    TEST_ASSERT_EQUAL(18, offsetof(ImuSample, gz));
}

// ── 2. Scale factor tables ───────────────────────────────────────────────────

void test_accel_scale_g2(void) {
    TEST_ASSERT_FLOAT_WITHIN(1e-7, 2.0f / 32768.0f, accelScale(AccelRange::G2));
}

void test_accel_scale_g4(void) {
    TEST_ASSERT_FLOAT_WITHIN(1e-7, 4.0f / 32768.0f, accelScale(AccelRange::G4));
}

void test_accel_scale_g8(void) {
    TEST_ASSERT_FLOAT_WITHIN(1e-7, 8.0f / 32768.0f, accelScale(AccelRange::G8));
}

void test_accel_scale_g16(void) {
    TEST_ASSERT_FLOAT_WITHIN(1e-7, 16.0f / 32768.0f, accelScale(AccelRange::G16));
}

void test_gyro_scale_dps250(void) {
    TEST_ASSERT_FLOAT_WITHIN(1e-7, 250.0f / 32768.0f, gyroScale(GyroRange::DPS250));
}

void test_gyro_scale_dps2000(void) {
    TEST_ASSERT_FLOAT_WITHIN(1e-7, 2000.0f / 32768.0f, gyroScale(GyroRange::DPS2000));
}

// ── 3. Sample rate validation ────────────────────────────────────────────────

void test_valid_rates(void) {
    TEST_ASSERT_TRUE(isValidRate(50));
    TEST_ASSERT_TRUE(isValidRate(100));
    TEST_ASSERT_TRUE(isValidRate(200));
}

void test_invalid_rates(void) {
    TEST_ASSERT_FALSE(isValidRate(0));
    TEST_ASSERT_FALSE(isValidRate(75));
    TEST_ASSERT_FALSE(isValidRate(150));
    TEST_ASSERT_FALSE(isValidRate(201));
    TEST_ASSERT_FALSE(isValidRate(500));
}

void test_rate_divisor_50hz(void) {
    // 1000 / 50 - 1 = 19
    TEST_ASSERT_EQUAL(19, sampleRateDivisor(50));
}

void test_rate_divisor_100hz(void) {
    // 1000 / 100 - 1 = 9
    TEST_ASSERT_EQUAL(9, sampleRateDivisor(100));
}

void test_rate_divisor_200hz(void) {
    // 1000 / 200 - 1 = 4
    TEST_ASSERT_EQUAL(4, sampleRateDivisor(200));
}

void test_rate_divisor_invalid_returns_sentinel(void) {
    TEST_ASSERT_EQUAL(0xFFFF, sampleRateDivisor(75));
    TEST_ASSERT_EQUAL(0xFFFF, sampleRateDivisor(0));
}

// ── 4. FIFO overflow detection ───────────────────────────────────────────────

void test_fifo_not_overflowed_under_capacity(void) {
    TEST_ASSERT_FALSE(fifoOverflowed(0));
    TEST_ASSERT_FALSE(fifoOverflowed(120));    // 10 samples
    TEST_ASSERT_FALSE(fifoOverflowed(504));    // 42 samples
    TEST_ASSERT_FALSE(fifoOverflowed(511));    // just under
}

void test_fifo_overflowed_at_capacity(void) {
    TEST_ASSERT_TRUE(fifoOverflowed(512));     // exactly capacity
    TEST_ASSERT_TRUE(fifoOverflowed(1024));    // way over
}

// ── 5. FIFO byte parsing (big-endian int16) ──────────────────────────────────

void test_parse_fifo_sample_positive_values(void) {
    uint8_t raw[12] = {
        0x00, 0x64,   // ax = 100
        0x00, 0xC8,   // ay = 200
        0x01, 0x2C,   // az = 300
        0x00, 0x0A,   // gx = 10
        0x00, 0x14,   // gy = 20
        0x00, 0x1E,   // gz = 30
    };
    ImuSample s{};
    parseFifoSample(raw, s);
    TEST_ASSERT_EQUAL(100,  s.ax);
    TEST_ASSERT_EQUAL(200,  s.ay);
    TEST_ASSERT_EQUAL(300,  s.az);
    TEST_ASSERT_EQUAL(10,   s.gx);
    TEST_ASSERT_EQUAL(20,   s.gy);
    TEST_ASSERT_EQUAL(30,   s.gz);
}

void test_parse_fifo_sample_negative_values(void) {
    uint8_t raw[12] = {
        0xFF, 0x9C,   // ax = -100
        0xFF, 0x38,   // ay = -200
        0xFE, 0xD4,   // az = -300
        0xFF, 0xF6,   // gx = -10
        0xFF, 0xEC,   // gy = -20
        0xFF, 0xE2,   // gz = -30
    };
    ImuSample s{};
    parseFifoSample(raw, s);
    TEST_ASSERT_EQUAL(-100, s.ax);
    TEST_ASSERT_EQUAL(-200, s.ay);
    TEST_ASSERT_EQUAL(-300, s.az);
    TEST_ASSERT_EQUAL(-10,  s.gx);
    TEST_ASSERT_EQUAL(-20,  s.gy);
    TEST_ASSERT_EQUAL(-30,  s.gz);
}

void test_parse_fifo_sample_max_min(void) {
    uint8_t raw[12] = {
        0x7F, 0xFF,   // ax = 32767 (max int16)
        0x80, 0x00,   // ay = -32768 (min int16)
        0x00, 0x00,   // az = 0
        0x7F, 0xFF,   // gx = 32767
        0x80, 0x00,   // gy = -32768
        0x00, 0x00,   // gz = 0
    };
    ImuSample s{};
    parseFifoSample(raw, s);
    TEST_ASSERT_EQUAL(32767,  s.ax);
    TEST_ASSERT_EQUAL(-32768, s.ay);
    TEST_ASSERT_EQUAL(0,      s.az);
    TEST_ASSERT_EQUAL(32767,  s.gx);
    TEST_ASSERT_EQUAL(-32768, s.gy);
    TEST_ASSERT_EQUAL(0,      s.gz);
}

// ── 6. Timestamp interpolation ───────────────────────────────────────────────
// Critical for clock sync (#7): each sample in a batch must get a distinct
// timestamp reflecting when it was actually captured, not the drain time.

void test_interpolate_timestamp_single_sample(void) {
    // 1 sample: timestamp = drain time (no offset)
    TEST_ASSERT_EQUAL(1000000u, interpolateTimestamp(1000000u, 100, 1, 0));
}

void test_interpolate_timestamp_oldest_in_batch(void) {
    // 5 samples at 100 Hz (interval = 10000 us), oldest (index 0):
    // drain - (5-1-0) * 10000 = drain - 40000
    const uint32_t drain = 5000000u;
    const uint32_t expected = drain - 40000u;
    TEST_ASSERT_EQUAL(expected, interpolateTimestamp(drain, 100, 5, 0));
}

void test_interpolate_timestamp_newest_in_batch(void) {
    // 5 samples at 100 Hz, newest (index 4):
    // drain - (5-1-4) * 10000 = drain - 0 = drain
    const uint32_t drain = 5000000u;
    TEST_ASSERT_EQUAL(drain, interpolateTimestamp(drain, 100, 5, 4));
}

void test_interpolate_timestamp_middle_in_batch(void) {
    // 5 samples at 100 Hz, middle (index 2):
    // drain - (5-1-2) * 10000 = drain - 20000
    const uint32_t drain = 5000000u;
    const uint32_t expected = drain - 20000u;
    TEST_ASSERT_EQUAL(expected, interpolateTimestamp(drain, 100, 5, 2));
}

void test_interpolate_timestamp_200hz_interval(void) {
    // 3 samples at 200 Hz (interval = 5000 us), oldest:
    // drain - (3-1-0) * 5000 = drain - 10000
    const uint32_t drain = 2000000u;
    TEST_ASSERT_EQUAL(drain - 10000u, interpolateTimestamp(drain, 200, 3, 0));
}

void test_interpolate_timestamp_50hz_interval(void) {
    // 2 samples at 50 Hz (interval = 20000 us), oldest:
    // drain - (2-1-0) * 20000 = drain - 20000
    const uint32_t drain = 1000000u;
    TEST_ASSERT_EQUAL(drain - 20000u, interpolateTimestamp(drain, 50, 2, 0));
}

void test_interpolate_timestamps_are_distinct(void) {
    // The bug we fixed: all samples in a batch must NOT have the same timestamp
    const uint32_t drain = 5000000u;
    const uint32_t t0 = interpolateTimestamp(drain, 100, 5, 0);
    const uint32_t t1 = interpolateTimestamp(drain, 100, 5, 1);
    const uint32_t t2 = interpolateTimestamp(drain, 100, 5, 2);
    const uint32_t t3 = interpolateTimestamp(drain, 100, 5, 3);
    const uint32_t t4 = interpolateTimestamp(drain, 100, 5, 4);
    TEST_ASSERT_TRUE(t0 < t1);
    TEST_ASSERT_TRUE(t1 < t2);
    TEST_ASSERT_TRUE(t2 < t3);
    TEST_ASSERT_TRUE(t3 < t4);
}

void test_interpolate_timestamps_spacing_matches_rate(void) {
    // Consecutive samples should be exactly 1/rate_hz apart
    const uint32_t drain = 5000000u;
    const uint32_t t0 = interpolateTimestamp(drain, 100, 5, 0);
    const uint32_t t1 = interpolateTimestamp(drain, 100, 5, 1);
    TEST_ASSERT_EQUAL(10000u, t1 - t0);   // 1/100 Hz = 10000 us
}

// ── Runner ───────────────────────────────────────────────────────────────────

int main() {
    UNITY_BEGIN();

    // 1. Struct size
    RUN_TEST(test_imu_sample_is_20_bytes);
    RUN_TEST(test_imu_sample_field_offsets);

    // 2. Scale tables
    RUN_TEST(test_accel_scale_g2);
    RUN_TEST(test_accel_scale_g4);
    RUN_TEST(test_accel_scale_g8);
    RUN_TEST(test_accel_scale_g16);
    RUN_TEST(test_gyro_scale_dps250);
    RUN_TEST(test_gyro_scale_dps2000);

    // 3. Rate validation
    RUN_TEST(test_valid_rates);
    RUN_TEST(test_invalid_rates);
    RUN_TEST(test_rate_divisor_50hz);
    RUN_TEST(test_rate_divisor_100hz);
    RUN_TEST(test_rate_divisor_200hz);
    RUN_TEST(test_rate_divisor_invalid_returns_sentinel);

    // 4. FIFO overflow
    RUN_TEST(test_fifo_not_overflowed_under_capacity);
    RUN_TEST(test_fifo_overflowed_at_capacity);

    // 5. FIFO parsing
    RUN_TEST(test_parse_fifo_sample_positive_values);
    RUN_TEST(test_parse_fifo_sample_negative_values);
    RUN_TEST(test_parse_fifo_sample_max_min);

    // 6. Timestamp interpolation
    RUN_TEST(test_interpolate_timestamp_single_sample);
    RUN_TEST(test_interpolate_timestamp_oldest_in_batch);
    RUN_TEST(test_interpolate_timestamp_newest_in_batch);
    RUN_TEST(test_interpolate_timestamp_middle_in_batch);
    RUN_TEST(test_interpolate_timestamp_200hz_interval);
    RUN_TEST(test_interpolate_timestamp_50hz_interval);
    RUN_TEST(test_interpolate_timestamps_are_distinct);
    RUN_TEST(test_interpolate_timestamps_spacing_matches_rate);

    return UNITY_END();
}
